### PR TITLE
Introduce system property for controlling URI validation

### PR DESCRIPTION
**Pull request by geoand**
This is not meant to be used in production and is mainly
aimed at benchmarking.
The property is very much analogous to what Vert.x provides
with `vertx.disableHttpHeadersValidation`

cc @Sanne @johnaohara

### CODE REVIEW

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java [hunk 1]
```diff
@@ -141,15 +141,23 @@ public class VertxHttpRecorder {
 
     public static final String GET = "GET";
     private static final Handler<HttpServerRequest> ACTUAL_ROOT = new Handler<HttpServerRequest>() {
+
+        /** JVM system property that disables URI validation, don't use this in production. */
+        private static final String DISABLE_URI_VALIDATION_PROP_NAME = "vertx.disableURIValidation";
```
Comment (geoand) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:146]: I used the `vertx` prefix to keep it consistent with `vertx.disableHttpHeadersValidation`, but we can certainly 
use `quarkus` instead if we think it's more appropriate

**Review by Sanne — APPROVED** [review]
Interesting :)
Might be relevant for some work @franz1981 is doing today

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java (hunk 1 above)
Comment (geoand) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:146]: Well we are the ones doing this validation, not Vert.x 🙂

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java [hunk 2]
```diff
@@ -165,6 +173,19 @@ public void handle(HttpServerRequest httpServerRequest) {
...
         }
+
+        private boolean uriValid(HttpServerRequest httpServerRequest) {
+            if (DISABLE_URI_VALIDATION) {
+                return true;
+            }
+            try {
+                // we simply need to know if the URI is valid
+                new URI(httpServerRequest.uri());
+                return true;
```
Comment (gastaldi) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:184]: Technically a more accurate validation would be: 

```suggestion
            return uri.getScheme() != null && uri.getHost() != null;
```

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java (hunk 1 above)
Comment (gastaldi) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:146]: Sure, I meant that Vert.x should have validate that, not us 😉

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java [hunk 3]
```diff
@@ -141,15 +141,23 @@ public class VertxHttpRecorder {
...
     private static final Handler<HttpServerRequest> ACTUAL_ROOT = new Handler<HttpServerRequest>() {
+
+        /** JVM system property that disables URI validation, don't use this in production. */
+        private static final String DISABLE_URI_VALIDATION_PROP_NAME = "vertx.disableURIValidation";
+
+        /**
+         * Constant that disables HTTP headers validation, this is a constant so the JIT can eliminate validation code
+         * and save some allocations.
+         */
+        private final boolean DISABLE_URI_VALIDATION = Boolean.getBoolean(DISABLE_URI_VALIDATION_PROP_NAME);
```
Comment (franz1981) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:152]: must be static too in order to be treated as constant

**Review by franz1981 — CHANGES_REQUESTED** [review]

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java (hunk 2 above)
Comment (geoand) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:184]: Very true, but that's not what we do know, so best leave it as is 😎

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java (hunk 3 above)
Comment (gastaldi) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:152]: 👍🏻 Good catch!

Comment (geoand) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:152]: I thought of that, but decided against it so as to not have it captured during native image build

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java (hunk 2 above)
Comment (gastaldi) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:184]: That would prevent entries like `127.0.0.1` (without the `http://`) to be considered valid

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java (hunk 3 above)
Comment (franz1981) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:152]: Makes sense, but still the comment
```
this is a constant so the JIT can eliminate validation code and save some allocations.
```
is not valid anymore
Netty is using many system properties like these and I expect them to be correctly handled on native image (which knowledge I'm still lacking) - is it not the same?

Comment (geoand) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:152]: The constants we care about to not be captured in native image, we explicitly "reset" using a substitution.

We could of course do that here too, but it feels like too much work for such a niche feature

Comment (franz1981) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:152]: At this point let's treat this as a (non-JIT) final field that save some useless work (that's fine)

Comment (geoand) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:152]: Good point, however the handler is only created once :).

I did push an update to the comment

**Pull request merged**